README.rst
Python Firefox Sync client
##########################


This is a python client for Firefox Sync. Check it out with::

  $ python setup.py install
  $ python syncclient/main.py --help

For instance, if you want to get all passwords (encrypted) use the
`get_records` action:

.. code-block::

  $ python syncclient/main.py alexis@notmyidea.org $PASSWORD get_records passwords
  [u'{1c1e0eea-d9c2-4c59-b95e-4dbe0800639f}',
   u'{0a76ec08-ba7c-48b1-b026-1d65085f789e}',
   u'{7482b391-bf2f-4542-8ebd-27c4398487ff}',
   u'{37bc9298-ac49-c54e-a73d-d817434ed0b2}',
   u'{d5ff4718-d4a0-4703-b0af-7d1c79c3a099}']



CHANGELOG.rst
CHANGELOG
#########

This document describes changes between each past release.


0.9.0 (unreleased)
==================

- Nothing changed yet.


0.8.0 (2015-12-30)
==================

- Update to latest request_hawk API. (#24)


0.7.0 (2015-11-17)
==================

- Add support for trusted certificate pinning. (#21)


0.6.0 (2015-10-12)
==================

- Circumvent a limitation present in requests-hawk 0.2.0. (#17)
- Add support for sorting by oldest when retrieving records. (#18)


0.5.0 (2015-10-01)
==================

- Handle HTTP 304 Not Modified Firefox Sync respones (#14)


0.4.0 (2015-09-23)
==================

- Forward additionnal client keyword arguments to requests (#12)


0.3.0 (2015-09-21)
==================

- Separate SyncClient and TokenserverClient code.
- Add a parameter to configure the expiration of the TokenServer returned credentials.
- Add a parameter to create a SyncClient with already fetched TokenServer credentials.
- Handle TokenServer served from a prefixed path.


0.2.0 (2015-09-03)
==================

**Bug Fixes**

- Handle API Server URL version prefix.


0.1.0 (2015-09-03)
==================

**Initial version**

- A client to synchroneously call a Firefox Sync server.


tests/support.py
try:
    import unittest2 as unittest
except ImportError:
    import unittest  # NOQA

import mock


def patch(test_case, *to_mock):
    """Patch the given positional arguments, return the patched version.

    :param test_case:
        The test case to use, in order to unpatch at the end of the test.
    """
    patched = []
    for m in to_mock:
        p = mock.patch(m)
        patched.append(p.start())
        test_case.addCleanup(p.stop)
    return patched


tests/test_client.py
# -*- coding: utf-8 -*-
import mock
from hashlib import sha256
from requests.exceptions import HTTPError

from syncclient.client import (
    TokenserverClient, SyncClient, SyncClientError, TOKENSERVER_URL,
    get_browserid_assertion, encode_header
)
from .support import unittest, patch


class TokenserverClientTest(unittest.TestCase):
    def test_token_server_request_token_server_url(self):
        client = TokenserverClient("given_bid", "given_client_state")
        with mock.patch("syncclient.client.requests") as requests:
            client.get_hawk_credentials()
            requests.get.assert_called_with(
                "https://token.services.mozilla.com/1.0/sync/1.5",
                headers={
                    'Authorization': "BrowserID given_bid",
                    'X-Client-State': "given_client_state"
                }, params={}, verify=None)
            requests.get.return_value.raise_for_status.assert_called_with()
            requests.get.return_value.json.assert_called_with()

    def test_token_server_request_handle_duration_parameter(self):
        client = TokenserverClient("given_bid", "given_client_state")
        with mock.patch("syncclient.client.requests") as requests:
            client.get_hawk_credentials(duration=300)
            requests.get.assert_called_with(
                "https://token.services.mozilla.com/1.0/sync/1.5",
                headers={
                    'Authorization': "BrowserID given_bid",
                    'X-Client-State': "given_client_state"
                }, params={"duration": 300}, verify=None)
            requests.get.return_value.raise_for_status.assert_called_with()
            requests.get.return_value.json.assert_called_with()

    def test_token_server_client_can_be_pass_a_verify_parameter(self):
        client = TokenserverClient("given_bid", "given_client_state",
                                   verify='root-ca.crt')
        with mock.patch("syncclient.client.requests") as requests:
            client.get_hawk_credentials(duration=300)
            requests.get.assert_called_with(
                "https://token.services.mozilla.com/1.0/sync/1.5",
                headers={
                    'Authorization': "BrowserID given_bid",
                    'X-Client-State': "given_client_state"
                }, params={"duration": 300}, verify='root-ca.crt')


class SyncClientSetupTest(unittest.TestCase):
    def setUp(self):
        super(SyncClientSetupTest, self).setUp()
        patched = patch(self, 'syncclient.client.requests')
        self.requests = patched[0].request

    def test_wrong_syncclient_argument_raise_a_syncclienterror(self):
        try:
            SyncClient()
        except SyncClientError as e:
            self.assertEquals(
                e.args[0],
                "You should either provide a BID assertion and a client state "
                "or complete Sync credentials (uid, api_endpoint, hashalg, "
                "id, key)")
        else:
            self.fail("SyncClient should fail on missing parameters")

    def test_missing_syncclient_argument_raise_a_syncclienterror(self):
        try:
            SyncClient("bid_assertion_without_client_state")
        except SyncClientError as e:
            self.assertEquals(
                e.args[0],
                "You should either provide a BID assertion and a client state "
                "or complete Sync credentials (uid, api_endpoint, hashalg, "
                "id, key)")
        else:
            self.fail("SyncClient should fail on missing parameters")

    def test_missing_syncclient_credentials_raise_a_syncclienterror(self):
        try:
            SyncClient(uid="toto")
        except SyncClientError as e:
            self.assertEquals(
                e.args[0],
                "You should either provide a BID assertion and a client state "
                "or complete Sync credentials (uid, api_endpoint, hashalg, "
                "id, key)")
        else:
            self.fail("SyncClient should fail on missing parameters")

    def test_syncclient_can_be_setup_with_bid_and_client_state(self):
        with mock.patch("syncclient.client.TokenserverClient") as tokenserver:
            tokenserver.return_value.get_hawk_credentials.return_value = {
                "api_endpoint": "http://example.org/",
                "uid": "123456",
                "hashalg": "sha256",
                "id": "mon-id",
                "key": "I am not a secure key"
            }
            with mock.patch("syncclient.client.HawkAuth") as hawkauth:
                SyncClient("bid_assertion", "client_state")
                tokenserver.assert_called_with(
                    "bid_assertion", "client_state", TOKENSERVER_URL)
                tokenserver().get_hawk_credentials.assert_called_with()
                hawkauth.assert_called_with(algorithm="sha256",
                                            id="mon-id",
                                            key="I am not a secure key")

    def test_syncclient_can_be_setup_with_sync_credentials(self):
        credentials = {
            "api_endpoint": "http://example.org/",
            "uid": "123456",
            "hashalg": "sha256",
            "id": "mon-id",
            "key": "I am not a secure key"
        }
        with mock.patch("syncclient.client.TokenserverClient") as tokenserver:
            with mock.patch("syncclient.client.HawkAuth") as hawkauth:
                SyncClient(**credentials)
                tokenserver.assert_not_called()
                tokenserver().get_hawk_credentials.assert_not_called()
                hawkauth.assert_called_with(algorithm="sha256",
                                            id="mon-id",
                                            key="I am not a secure key")


class ClientRequestIssuanceTest(unittest.TestCase):
    def setUp(self):
        super(ClientRequestIssuanceTest, self).setUp()
        # Mock requests to avoid issuance of requests when we start the client.
        patched = patch(self, 'syncclient.client.requests')
        self.requests = patched[0].request
        self.requests.return_value.status_code = 200

    def _get_client(self, api_endpoint='http://example.org/', **kwargs):
        client = SyncClient("bid_assertion", "client_state", **kwargs)
        client.api_endpoint = api_endpoint
        client.auth = mock.sentinel.auth
        return client

    def test_client_add_serverurl_to_requests(self):
        client = self._get_client()
        client._request('get', '/test')

        self.requests.assert_called_with(
            'get', 'http://example.org/test',
            auth=client.auth, verify=None)

    def test_client_add_serverurl_with_prefix_to_requests(self):
        client = self._get_client("https://example.org/v1/")
        client._request('get', '/test')

        self.requests.assert_called_with(
            'get', 'https://example.org/v1/test',
            auth=client.auth, verify=None)

    def test_request_raise_on_error(self):
        # Patch requests to raise an exception.
        resp = mock.MagicMock()
        resp.raise_for_status.side_effect = Exception
        self.requests.return_value = resp

        client = self._get_client()
        self.assertRaises(Exception, client._request, 'get', '/')

    def test_client_add_the_verify_parameter_to_requests(self):
        client = self._get_client("https://example.org/v1/",
                                  verify='root-ca.crt')
        client._request('get', '/test')

        self.requests.assert_called_with(
            'get', 'https://example.org/v1/test',
            auth=client.auth, verify='root-ca.crt')


class ClientAuthenticationTest(unittest.TestCase):
    def setUp(self):
        super(ClientAuthenticationTest, self).setUp()
        patched = patch(self, 'syncclient.client.requests',
                        'syncclient.client.HawkAuth')
        self.requests = patched[0]
        self.hawk_auth = patched[1]

    def test_authenticate_requests_the_tokenserver_with_proper_headers(self):
        SyncClient(u"bid_assertion", "client_state")
        self.requests.get.assert_called_with(
            'https://token.services.mozilla.com/1.0/sync/1.5',
            headers={
                'X-Client-State': 'client_state',
                'Authorization': 'BrowserID bid_assertion'
            }, params={}, verify=None)

    def test_error_with_tokenserver_is_raised(self):
        resp = mock.MagicMock()
        resp.raise_for_status.side_effect = Exception
        self.requests.get.return_value = resp
        self.assertRaises(Exception, SyncClient, "bid_assertion",
                          "client_state")

    def test_credentials_from_tokenserver_are_passed_to_hawkauth(self):
        resp = mock.MagicMock()
        resp.json.return_value = {
            'hashalg': mock.sentinel.hashalg,
            'id': mock.sentinel.id,
            'key': mock.sentinel.key,
            'uid': mock.sentinel.uid,
            'api_endpoint': mock.sentinel.api_endpoint
        }
        self.requests.get.return_value = resp
        client = SyncClient("bid_assertion", "client_state")

        self.hawk_auth.assert_called_with(algorithm=mock.sentinel.hashalg,
                                          id=mock.sentinel.id,
                                          key=mock.sentinel.key)

        assert client.user_id == mock.sentinel.uid
        assert client.api_endpoint == mock.sentinel.api_endpoint


class BrowserIDAssertionTest(unittest.TestCase):

    @mock.patch('syncclient.client.FxAClient')
    @mock.patch('syncclient.client.hexlify')
    def test_trade_works_as_expected(self, hexlify, fxa_client):
        # mock the calls to PyFxA.
        fake_keyB = "fake key b"
        session = mock.MagicMock()
        session.fetch_keys.return_value = None, fake_keyB
        fxa_client().login.return_value = session
        get_browserid_assertion('login', 'password')

        digest = sha256(fake_keyB.encode('utf-8')).digest()[0:16]
        hexlify.return_value = mock.sentinel.hexlified
        hexlify.assert_called_with(digest)


class ClientHTTPCallsTest(unittest.TestCase):
    def setUp(self):
        super(ClientHTTPCallsTest, self).setUp()
        self.client = SyncClient(
            hashalg=mock.sentinel.hashalg,
            id=mock.sentinel.id,
            key=mock.sentinel.key,
            uid=mock.sentinel.uid,
            api_endpoint=mock.sentinel.api_endpoint
        )

        # Mock the request method of the client, since we'll use
        # it to make sure the correct requests are made.
        self.client._request = mock.MagicMock()

    def test_info_collection(self):
        self.client.info_collections()
        self.client._request.assert_called_with('get', '/info/collections')

    def test_info_collection_can_receive_requests_parameters(self):
        self.client.info_collections(headers=mock.sentinel.headers)
        self.client._request.assert_called_with('get', '/info/collections',
                                                headers=mock.sentinel.headers)

    def test_info_quota(self):
        self.client.info_quota()
        self.client._request.assert_called_with('get', '/info/quota')

    def test_info_quota_can_receive_requests_parameters(self):
        self.client.info_quota(headers=mock.sentinel.headers)
        self.client._request.assert_called_with('get', '/info/quota',
                                                headers=mock.sentinel.headers)

    def test_collection_usage(self):
        self.client.get_collection_usage()
        self.client._request.assert_called_with(
            'get', '/info/collection_usage')

    def test_collection_usage_can_receive_requests_parameters(self):
        self.client.get_collection_usage(headers=mock.sentinel.headers)
        self.client._request.assert_called_with(
            'get', '/info/collection_usage', headers=mock.sentinel.headers)

    def test_collection_counts(self):
        self.client.get_collection_counts()
        self.client._request.assert_called_with(
            'get', '/info/collection_counts')

    def test_collection_counts_can_receive_requests_parameters(self):
        self.client.get_collection_counts(headers=mock.sentinel.headers)
        self.client._request.assert_called_with(
            'get', '/info/collection_counts', headers=mock.sentinel.headers)

    def test_delete_all_records(self):
        self.client.delete_all_records()
        self.client._request.assert_called_with(
            'delete', '/')

    def test_delete_all_records_can_receive_requests_parameters(self):
        self.client.delete_all_records(headers=mock.sentinel.headers)
        self.client._request.assert_called_with(
            'delete', '/', headers=mock.sentinel.headers)

    def test_get_records_sets_full_by_default(self):
        self.client.get_records('mycollection')
        self.client._request.assert_called_with(
            'get', '/storage/mycollection',
            params={'full': True})

    def test_get_records_lowers_the_collection_name(self):
        self.client.get_records('myCollection')
        self.client._request.assert_called_with(
            'get', '/storage/mycollection',
            params={'full': True})

    def test_get_records_handles_ids(self):
        self.client.get_records('myCollection', ids=(1, 3))
        self.client._request.assert_called_with(
            'get', '/storage/mycollection',
            params={'full': True, 'ids': '1,3'})

    def test_get_records_handles_full(self):
        self.client.get_records('mycollection', full=False)
        self.client._request.assert_called_with(
            'get', '/storage/mycollection',
            params={})

    def test_get_records_handles_newer(self):
        self.client.get_records('mycollection', newer='newer')
        self.client._request.assert_called_with(
            'get', '/storage/mycollection',
            params={'newer': 'newer', 'full': True})

    def test_get_records_handles_limit(self):
        self.client.get_records('mycollection', limit='limit')
        self.client._request.assert_called_with(
            'get', '/storage/mycollection',
            params={'limit': 'limit', 'full': True})

    def test_get_records_handles_offset(self):
        self.client.get_records('mycollection', offset='offset')
        self.client._request.assert_called_with(
            'get', '/storage/mycollection',
            params={'offset': 'offset', 'full': True})

    def test_get_records_handles_sort_by_newest(self):
        self.client.get_records('mycollection', sort='newest')
        self.client._request.assert_called_with(
            'get', '/storage/mycollection',
            params={'sort': 'newest', 'full': True})

    def test_get_records_handles_sort_by_oldest(self):
        self.client.get_records('mycollection', sort='oldest')
        self.client._request.assert_called_with(
            'get', '/storage/mycollection',
            params={'sort': 'oldest', 'full': True})

    def test_get_records_handles_sort_by_index(self):
        self.client.get_records('mycollection', sort='index')
        self.client._request.assert_called_with(
            'get', '/storage/mycollection',
            params={'sort': 'index', 'full': True})

    def test_get_records_ignore_sort_by_invalid(self):
        self.client.get_records('mycollection', sort='invalid')
        self.client._request.assert_called_with(
            'get', '/storage/mycollection',
            params={'full': True})

    def test_get_record(self):
        self.client.get_record('myCollection', 1234)
        self.client._request.assert_called_with(
            'get', '/storage/mycollection/1234')

    def test_get_record_can_receive_requests_parameters(self):
        self.client.get_record('myCollection', 1234,
                               headers=mock.sentinel.headers)
        self.client._request.assert_called_with(
            'get', '/storage/mycollection/1234',
            headers=mock.sentinel.headers)

    def test_delete_record(self):
        self.client.delete_record('myCollection', 1234)
        self.client._request.assert_called_with(
            'delete', '/storage/mycollection/1234')

    def test_delete_record_can_receive_requests_parameters(self):
        self.client.delete_record('myCollection', 1234,
                                  headers=mock.sentinel.headers)
        self.client._request.assert_called_with(
            'delete', '/storage/mycollection/1234',
            headers=mock.sentinel.headers)

    def test_put_record(self):
        record = {'id': 1234, 'foo': 'bar'}
        self.client.put_record('myCollection', record)
        self.client._request.assert_called_with(
            'put', '/storage/mycollection/1234',
            data='{"foo": "bar"}',
            headers={'Content-Type': 'application/json; charset=utf-8'})

    def test_put_record_handle_json_string_parameter(self):
        record = '{"id": 1234, "foo": "bar"}'
        self.client.put_record('myCollection', record)
        self.client._request.assert_called_with(
            'put', '/storage/mycollection/1234',
            data='{"foo": "bar"}',
            headers={'Content-Type': 'application/json; charset=utf-8'})

    def test_put_record_can_receive_requests_parameters(self):
        record = {'id': 1234, 'foo': 'bar'}
        self.client.put_record('myCollection', record,
                               headers={'Sentinel': 'true'})
        self.client._request.assert_called_with(
            'put', '/storage/mycollection/1234',
            data='{"foo": "bar"}',
            headers={'Content-Type': 'application/json; charset=utf-8',
                     'Sentinel': 'true'})

    def test_put_record_doesnt_modify_the_passed_object(self):
        record = {'id': 1234, 'foo': 'bar'}
        self.client.put_record('myCollection', record)
        assert 'id' in record.keys()

    def test_post_records(self):
        # For now, this does nothing.
        records = [{'id': idx, 'foo': 'foo'} for idx in range(1, 10)]
        self.client.post_records("myCollection", records)


class HandleSyncRequestsResponseTest(unittest.TestCase):
    def setUp(self):
        super(HandleSyncRequestsResponseTest, self).setUp()
        self.client = SyncClient(
            hashalg=mock.sentinel.hashalg,
            id=mock.sentinel.id,
            key=mock.sentinel.key,
            uid=mock.sentinel.uid,
            api_endpoint="http://sync.services.mozilla.com/"
        )

    def test_get_record_can_handle_empty_response(self):
        with mock.patch("syncclient.client.requests.request") as request:
            response = mock.MagicMock()
            response.status_code = 304
            response.response = "Not Modified"
            response.url = "http://www.example.com/"
            request.return_value = response
            self.assertRaises(HTTPError,
                              self.client.get_record, 'myCollection', 1234)


class EncodeHeaderTest(unittest.TestCase):
    def test_encode_str_return_str(self):
        value = 'Toto'
        self.assertEqual(type(value), str)
        value = encode_header(value)
        self.assertEqual(type(value), str)

    def test_returns_a_string_if_passed_bytes(self):
        entry = 'Toto'.encode('utf-8')
        value = encode_header(entry)
        self.assertEqual(type(value), str)

    def test_returns_a_string_if_passed_unicode(self):
        entry = u'Rémy'
        value = encode_header(entry)
        self.assertEqual(type(value), str)


syncclient/main.py
import argparse
from client import SyncClient, get_browserid_assertion
from pprint import pprint


def main():
    parser = argparse.ArgumentParser(
        description="""CLI to interact with Firefox Sync""",
        formatter_class=argparse.ArgumentDefaultsHelpFormatter
    )
    parser.add_argument(dest='login',
                        help='Firefox Accounts login.')
    parser.add_argument(dest='password',
                        help='Firefox Accounts password.')
    parser.add_argument(dest='action', help='The action to be executed',
                        default='info_collections', nargs='?',
                        choices=[m for m in dir(SyncClient)
                                 if not m.startswith('_')])

    args, extra = parser.parse_known_args()

    bid_assertion_args = get_browserid_assertion(args.login, args.password)
    client = SyncClient(*bid_assertion_args)
    pprint(getattr(client, args.action)(*extra))

if __name__ == '__main__':
    main()


syncclient/client.py
from hashlib import sha256
from binascii import hexlify
import json
import six
import sys

import requests
from requests_hawk import HawkAuth
from fxa.core import Client as FxAClient

# This is a proof of concept, in python, to get some data of some collections.
# The data stays encrypted and because we don't have the keys to decrypt it
# it just stays like that for now. The goal is simply to prove that it's
# possible to get the data out of the API"""

TOKENSERVER_URL = "https://token.services.mozilla.com/"
FXA_SERVER_URL = "https://api.accounts.firefox.com"


def encode_header(value):
    if isinstance(value, str):
        return value
    # Python3, it must be bytes
    if sys.version_info[0] > 2:  # pragma: no cover
        return value.decode('utf-8')
    # Python2, it must be unicode
    else:  # pragma: no cover
        return value.encode('utf-8')


def get_browserid_assertion(login, password, fxa_server_url=FXA_SERVER_URL,
                            tokenserver_url=TOKENSERVER_URL):
    """Trade a user and password for a BrowserID assertion and the client
    state.
    """
    client = FxAClient(server_url=fxa_server_url)
    session = client.login(login, password, keys=True)
    bid_assertion = session.get_identity_assertion(tokenserver_url)
    _, keyB = session.fetch_keys()
    if isinstance(keyB, six.text_type):  # pragma: no cover
        keyB = keyB.encode('utf-8')
    return bid_assertion, hexlify(sha256(keyB).digest()[0:16])


class SyncClientError(Exception):
    """An error occured in SyncClient."""


class TokenserverClient(object):
    """Client for the Firefox Sync Token Server.
    """
    def __init__(self, bid_assertion, client_state,
                 server_url=TOKENSERVER_URL, verify=None):
        self.bid_assertion = bid_assertion
        self.client_state = client_state
        self.server_url = server_url
        self.verify = verify

    def get_hawk_credentials(self, duration=None):
        """Asks for new temporary token given a BrowserID assertion"""
        authorization = 'BrowserID %s' % encode_header(self.bid_assertion)
        headers = {
            'Authorization': authorization,
            'X-Client-State': self.client_state
        }
        params = {}

        if duration is not None:
            params['duration'] = int(duration)

        url = self.server_url.rstrip('/') + '/1.0/sync/1.5'
        raw_resp = requests.get(url, headers=headers, params=params,
                                verify=self.verify)
        raw_resp.raise_for_status()
        return raw_resp.json()


class SyncClient(object):
    """Client for the Firefox Sync server.
    """

    def __init__(self, bid_assertion=None, client_state=None,
                 tokenserver_url=TOKENSERVER_URL, verify=None, **credentials):

        if bid_assertion is not None and client_state is not None:
            ts_client = TokenserverClient(bid_assertion, client_state,
                                          tokenserver_url)
            credentials = ts_client.get_hawk_credentials()

        else:
            # Make sure if the user wants to use credentials that they
            # give all the needed information.
            credentials_complete = set(credentials.keys()).issuperset({
                'uid', 'api_endpoint', 'hashalg', 'id', 'key'})

            if not credentials_complete:
                raise SyncClientError(
                    "You should either provide a BID assertion and a client "
                    "state or complete Sync credentials (uid, api_endpoint, "
                    "hashalg, id, key)")

        self.user_id = credentials['uid']
        self.api_endpoint = credentials['api_endpoint']
        self.auth = HawkAuth(algorithm=credentials['hashalg'],
                             id=credentials['id'],
                             key=credentials['key'])
        self.verify = verify

    def _request(self, method, url, **kwargs):
        """Utility to request an endpoint with the correct authentication
        setup, raises on errors and returns the JSON.

        """
        url = self.api_endpoint.rstrip('/') + '/' + url.lstrip('/')
        kwargs.setdefault('verify', self.verify)
        self.raw_resp = requests.request(method, url, auth=self.auth, **kwargs)
        self.raw_resp.raise_for_status()

        if self.raw_resp.status_code == 304:
            http_error_msg = '%s Client Error: %s for url: %s' % (
                self.raw_resp.status_code,
                self.raw_resp.reason,
                self.raw_resp.url)
            raise requests.exceptions.HTTPError(http_error_msg,
                                                response=self.raw_resp)
        return self.raw_resp.json()

    def info_collections(self, **kwargs):
        """
        Returns an object mapping collection names associated with the account
        to the last-modified time for each collection.

        The server may allow requests to this endpoint to be authenticated
        with an expired token, so that clients can check for server-side
        changes before fetching an updated token from the Token Server.
        """
        return self._request('get', '/info/collections', **kwargs)

    def info_quota(self, **kwargs):
        """
        Returns a two-item list giving the user's current usage and quota
        (in KB). The second item will be null if the server does not enforce
        quotas.

        Note that usage numbers may be approximate.
        """
        return self._request('get', '/info/quota', **kwargs)

    def get_collection_usage(self, **kwargs):
        """
        Returns an object mapping collection names associated with the account
        to the data volume used for each collection (in KB).

        Note that these results may be very expensive as it calculates more
        detailed and accurate usage information than the info_quota method.
        """
        return self._request('get', '/info/collection_usage', **kwargs)

    def get_collection_counts(self, **kwargs):
        """
        Returns an object mapping collection names associated with the
        account to the total number of items in each collection.
        """
        return self._request('get', '/info/collection_counts', **kwargs)

    def delete_all_records(self, **kwargs):
        """Deletes all records for the user."""
        return self._request('delete', '/', **kwargs)

    def get_records(self, collection, full=True, ids=None, newer=None,
                    limit=None, offset=None, sort=None, **kwargs):
        """
        Returns a list of the BSOs contained in a collection. For example:

        >>> ["GXS58IDC_12", "GXS58IDC_13", "GXS58IDC_15"]

        By default only the BSO ids are returned, but full objects can be
        requested using the full parameter. If the collection does not exist,
        an empty list is returned.

        :param ids:
            a comma-separated list of ids. Only objects whose id is in
            this list will be returned. A maximum of 100 ids may be provided.

        :param newer:
            a timestamp. Only objects whose last-modified time is strictly
            greater than this value will be returned.

        :param full:
            any value. If provided then the response will be a list of full
            BSO objects rather than a list of ids.

        :param limit:
            a positive integer. At most that many objects will be returned.
            If more than that many objects matched the query,
            an X-Weave-Next-Offset header will be returned.

        :param offset:
            a string, as returned in the X-Weave-Next-Offset header of a
            previous request using the limit parameter.

        :param sort:
            sorts the output:
            "newest" - orders by last-modified time, largest first
            "index" - orders by the sortindex, highest weight first
            "oldest" - orders by last-modified time, oldest first
        """
        params = kwargs.pop('params', {})
        if full:
            params['full'] = True
        if ids is not None:
            params['ids'] = ','.join(map(str, ids))
        if newer is not None:
            params['newer'] = newer
        if limit is not None:
            params['limit'] = limit
        if offset is not None:
            params['offset'] = offset
        if sort is not None and sort in ('newest', 'index', 'oldest'):
            params['sort'] = sort

        return self._request('get', '/storage/%s' % collection.lower(),
                             params=params, **kwargs)

    def get_record(self, collection, record_id, **kwargs):
        """Returns the BSO in the collection corresponding to the requested id.
        """
        return self._request('get', '/storage/%s/%s' % (collection.lower(),
                                                        record_id), **kwargs)

    def delete_record(self, collection, record_id, **kwargs):
        """Deletes the BSO at the given location.
        """
        return self._request('delete', '/storage/%s/%s' % (
            collection.lower(), record_id), **kwargs)

    def put_record(self, collection, record, **kwargs):
        """
        Creates or updates a specific BSO within a collection.
        The passed record must be a python object containing new data for the
        BSO.

        If the target BSO already exists then it will be updated with the
        data from the request body. Fields that are not provided will not be
        overwritten, so it is possible to e.g. update the ttl field of a
        BSO without re-submitting its payload. Fields that are explicitly set
        to null in the request body will be set to their default value by the
        server.

        If the target BSO does not exist, then fields that are not provided in
        the python object will be set to their default value by the server.

        Successful responses will return the new last-modified time for the
        collection.

        Note that the server may impose a limit on the amount of data
        submitted for storage in a single BSO.
        """
        # XXX: Workaround until request-hawk supports the json parameter. (#17)
        if isinstance(record, six.string_types):
            record = json.loads(record)
        record = record.copy()
        record_id = record.pop('id')
        headers = {}
        if 'headers' in kwargs:
            headers = kwargs.pop('headers')

        headers['Content-Type'] = 'application/json; charset=utf-8'

        return self._request('put', '/storage/%s/%s' % (
            collection.lower(), record_id), data=json.dumps(record),
            headers=headers, **kwargs)

    def post_records(self, collection, records, **kwargs):
        """
        Takes a list of BSOs in the request body and iterates over them,
        effectively doing a series of individual PUTs with the same timestamp.

        Each BSO record must include an "id" field, and the corresponding BSO
        will be created or updated according to the semantics of a PUT request
        targeting that specific record.

        In particular, this means that fields not provided will not be
        overwritten on BSOs that already exist.

        Successful responses will contain a JSON object with details of
        success or failure for each BSO. It will have the following keys:

            modified: the new last-modified time for the updated items.
            success: a (possibly empty) list of ids of BSOs that were
                     successfully stored.
            failed: a (possibly empty) object whose keys are the ids of BSOs
                    that were not stored successfully, and whose values are
                    lists of strings describing possible reasons for the
                    failure.

        For example:

        {
         "modified": 1233702554.25,
         "success": ["GXS58IDC_12", "GXS58IDC_13", "GXS58IDC_15",
                     "GXS58IDC_16", "GXS58IDC_18", "GXS58IDC_19"],
         "failed": {"GXS58IDC_11": ["invalid ttl"],
                    "GXS58IDC_14": ["invalid sortindex"]}
        }

        Posted BSOs whose ids do not appear in either "success" or "failed"
        should be treated as having failed for an unspecified reason.

        Note that the server may impose a limit on the total amount of data
        included in the request, and/or may decline to process more than a
        certain number of BSOs in a single request. The default limit on the
        number of BSOs per request is 100.
        """
        pass
